crates/rust-analyzer/src/global_state.rs

```diff
@@ -197,37 +197,51 @@ impl GlobalState {
             // We need to fix up the changed events a bit, if we have a create or modify for a file
             // id that is followed by a delete we actually no longer observe the file text from the
             // create or modify which may cause problems later on
+            let mut collapsed_create_delete = false;
             changed_files.dedup_by(|a, b| {
                 use vfs::ChangeKind::*;
 
+                let has_collapsed_create_delete = mem::replace(&mut collapsed_create_delete, false);
+
                 if a.file_id != b.file_id {
                     return false;
                 }
 
-                match (a.change_kind, b.change_kind) {
+                // true => delete the second element (a), we swap them here as they are inverted by dedup_by
+                match (b.change_kind, a.change_kind) {
                     // duplicate can be merged
                     (Create, Create) | (Modify, Modify) | (Delete, Delete) => true,
                     // just leave the create, modify is irrelevant
-                    (Create, Modify) => {
-                        std::mem::swap(a, b);
+                    (Create, Modify) => true,
+                    // modify becomes irrelevant if the file is deleted
+                    (Modify, Delete) => {
+                        mem::swap(a, b);
+                        true
+                    }
+                    // Remove the create message, and in the following loop, also remove the delete
+                    (Create, Delete) => {
+                        collapsed_create_delete = true;
+                        b.change_kind = Delete;
+                        true
+                    }
+                    // trailing delete from earlier
+                    (Delete, Create | Modify) if has_collapsed_create_delete => {
+                        b.change_kind = Create;
                         true
                     }
-                    // modify becomes irrelevant if the file is deleted
-                    (Modify, Delete) => true,
-                    // we should fully remove this occurrence,
-                    // but leaving just a delete works as well
-                    (Create, Delete) => true,
                     // this is equivalent to a modify
                     (Delete, Create) => {
-                        a.change_kind = Modify;
+                        b.change_kind = Modify;
                         true
                     }
                     // can't really occur
                     (Modify, Create) => false,
                     (Delete, Modify) => false,
                 }
             });
-
+            if collapsed_create_delete {
+                changed_files.pop();
```

Comment:
I've only skimmed the changes, but does this work for `[Create(f1), Delete(f1), Create(f2), Delete(f2)]`?

And generally, I don't think `dedup_by` guarantees anything about the order of calls. Maybe we should "just" keep a map from `FileId` to `Change`, go over the input once and update the map, without sorting it first? It might even be (asymptotically) faster.

Reply:
It should work, because the flag is only `true` at this point if it was set to `true` during the last call to the closure (the beginning of the closure resets it to `false`).

I agree that the code is hard to follow though, and `dedup_by` is probably not helping with that.

Reply:
But on my example, this results in `[Delete(f1)]`, which is probably not desirable. `collapsed_create_delete` will only cause the last `Delete` to be dropped.

Reply:
A simple map might be a better choice here indeed